File: oauth2/src/main/java/chatox/oauth2/config/RabbitMQConfig.java
package chatox.oauth2.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.amqp.core.Binding;
import org.springframework.amqp.core.BindingBuilder;
import org.springframework.amqp.core.Queue;
import org.springframework.amqp.core.TopicExchange;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class RabbitMQConfig {

    @Bean
    @Autowired
    public RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory,
                                         Jackson2JsonMessageConverter jackson2JsonMessageConverter) {
        var rabbitTemplate = new RabbitTemplate(connectionFactory);
        rabbitTemplate.setMessageConverter(jackson2JsonMessageConverter);
        return rabbitTemplate;
    }

    @Bean
    @Autowired
    public Jackson2JsonMessageConverter jackson2JsonMessageConverter(ObjectMapper objectMapper) {
        return new Jackson2JsonMessageConverter(objectMapper);
    }

    @Bean
    public TopicExchange globalBanEvents() {
        return new TopicExchange("global.ban.events");
    }

    @Bean
    public Queue globalBanCreatedQueue() {
        return new Queue("oauth2_service_global_ban_created");
    }

    @Bean
    public Queue globalBanUpdatedQueue() {
        return new Queue("oauth2_service_global_ban_updated");
    }

    @Bean
    public Binding globalBanCreatedBinding() {
        return BindingBuilder
                .bind(globalBanCreatedQueue())
                .to(globalBanEvents())
                .with("global.ban.created.#");
    }

    @Bean
    public Binding globalBanUpdatedBinding() {
        return BindingBuilder
                .bind(globalBanUpdatedQueue())
                .to(globalBanEvents())
                .with("global.ban.updated.#");
    }
}

File: user/src/main/kotlin/chatox/user/config/RabbitMqConfig.kt
package chatox.user.config

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.amqp.core.Binding
import org.springframework.amqp.core.BindingBuilder
import org.springframework.amqp.core.Queue
import org.springframework.amqp.core.TopicExchange
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class RabbitMqConfig {
    @Bean
    fun userEvents() = TopicExchange("user.events")

    @Bean
    fun accountDeletedQueue() = Queue("user_service_account_deleted")

    @Bean
    fun userEventsBinding(): Binding = BindingBuilder
            .bind(accountDeletedQueue())
            .to(userEvents())
            .with("account.deleted.#")

    @Bean
    fun uploadEvents() = TopicExchange("upload.events")

    @Bean
    fun imageCreatedQueue() = Queue("user_service_image_created")

    @Bean
    fun uploadEventsBinding(): Binding = BindingBuilder
            .bind(imageCreatedQueue())
            .to(uploadEvents())
            .with("upload.image.created.#")

    @Bean
    fun websocketEvents() = TopicExchange("websocket.events")

    @Bean
    fun userConnectedQueue() = Queue("user_service_user_connected")

    @Bean
    fun userDisconnectedQueue() = Queue("user_service_user_disconnected")

    @Bean
    fun userConnectedBinding(): Binding = BindingBuilder
            .bind(userConnectedQueue())
            .to(websocketEvents())
            .with("user.connected.#")

    @Bean
    fun userDisconnectedBinding(): Binding = BindingBuilder
            .bind(userDisconnectedQueue())
            .to(websocketEvents())
            .with("user.disconnected.#")

    @Bean
    fun emailUpdatedQueue() = Queue("user_service_email_updated")

    @Bean
    fun accountEvents() = TopicExchange("account.events")

    @Bean
    fun emailUpdatedBinding(): Binding = BindingBuilder
            .bind(emailUpdatedQueue())
            .to(accountEvents())
            .with("account.email.updated.#")

    @Autowired
    @Bean
    fun rabbitTemplate(connectionFactory: ConnectionFactory,
                       jackson2JsonMessageConverter: Jackson2JsonMessageConverter): RabbitTemplate {
        val rabbitTemplate = RabbitTemplate(connectionFactory)
        rabbitTemplate.messageConverter = jackson2JsonMessageConverter
        return rabbitTemplate
    }

    @Autowired
    @Bean
    fun jackson2JsonMessageConverter(objectMapper: ObjectMapper) = Jackson2JsonMessageConverter(objectMapper)
}
